src/nupic/ntypes/Dimensions.hpp

```diff
@@ -29,352 +29,149 @@
-   *
-   * A Dimensions object is invalid if any dimensions are 0 (except for "don't
-   * care")
-   *
-   * @returns
-   *       Whether Dimensions is "valid"
-   */
-  bool isValid() const;
-
-  /**
-   *
-   * @}
-   *
-   * @name Coordinate<->index mapping
-   *
-   * Coordinate<->index mapping is in lower-major order, i.e.
-   * for Region with dimensions `[2,3]`:
-   *
-   *     [0,0] -> index 0
-   *     [1,0] -> index 1
-   *     [0,1] -> index 2
-   *     [1,1] -> index 3
-   *     [0,2] -> index 4
-   *     [1,2] -> index 5
-   *
-   * @{
-   *
-   */
-
-  /**
-   * Convert a Coordinate to a linear index (in lower-major order).
-   *
-   * @param coordinate
-   *        The coordinate to be converted
-   *
-   * @returns
-   *        The linear index corresponding to @a coordinate
-   */
-  size_t getIndex(const Coordinate &coordinate) const;
-
-  /**
-   * Convert a linear index (in lower-major order) to a Coordinate.
-   *
-   * @param index
-   *        The linear index to be converted
-   *
-   * @returns
-   *        The Coordinate corresponding to @a index
-   */
-  Coordinate getCoordinate(const size_t index) const;
-
-  /**
-   *
-   * @}
-   *
-   * @name Misc
-   *
-   * @{
-   *
-   */
-
-  /**
-   *
-   * Convert the Dimensions object to string representation.
-   *
-   * In most cases, we want a human-readable string, but for
-   * serialization we want only the actual dimension values
-   *
-   * @param humanReadable
-   *        The default is @c true, make the string human-readable,
-   *        set to @c false for serialization
-   *
-   * @returns
-   *        The string representation of the Dimensions object
-   */
-  std::string toString(bool humanReadable = true) const;
-
-
-  /**
-   * The equivalence operator.
-   *
-   * Two Dimensions objects will be considered equivalent, if any of the
-   * following satisfies:
-   *
-   * * They have the same number of dimensions and the same size for every
-   * dimension.
-   * * Both of them have the size of 1 for everything dimensions, despite of
-   * how many dimensions they have, i.e. isOnes() returns @c true for both
-   * of them. Some linking scenarios require us to treat [1] equivalent to [1 1]
-   * etc.
-   *
-   * @param dims2
-   *        The Dimensions object being compared
-   *
-   * @returns
-   *        Whether this Dimensions object is equivalent to @a dims2.
+   * * Invalid     - Some dimension is 0.
    *
+   * There is a function to check for each of these states.
    */
-  bool operator==(const Dimensions &dims2) const;
-
-  /**
-   * The in-equivalence operator, the opposite of operator==().
-   *
-   * @param dims2
-   *        The Dimensions object being compared
-   *
-   * @returns
-   *        Whether this Dimensions object is not equivalent to @a dims2.
-   */
-  bool operator!=(const Dimensions &dims2) const;
+  bool isUnspecified() const { return(size() == 0); }
+  bool isDontcare()    const { return(size() == 1 && at(0) == 0); }
```

Comment:
nit, for readability, use `at(0) == DONTCARE` if you define it. (and would need to define it before the use)

Reply:
Oh, yes I will do that.